### PR TITLE
5532 – Maintain table selection when user returns to Data Explorer

### PR DESCRIPTION
**Pull request by bmarchena**
### Summary

This PR updates the application to preserve the user's selections in the Data Explorer when navigating away and back.

#### Tasks/Bug Numbers
 - Fixes AB[#5532](https://dev.azure.com/NYCPlanning/ITD/_sprints/taskboard/Open%20Source%20Engineering/ITD/2022/Q3/Sprint%20P?workitem=5532)

### Technical Explanation

Stores filter selections in Data Explorer in ```sessionStorage```. Added an action to the ```application.js``` controller to generate a URL with query params from values in ```sessionStorage``` (if any) when clicking on Data Explorer tab.

### CODE REVIEW

File: app/controllers/application.js [hunk 1]
```diff
@@ -1,6 +1,24 @@
...
+    const explorerURL = "/explorer/cities/NYC"
+    let newURL;
+    if (sessionStorage.length) {
+      let queryParams = new URLSearchParams()
+
+      for (let key in sessionStorage) {
+        queryParams.append(key, sessionStorage[key])
+      }
+
+      newURL = explorerURL + '?' + queryParams.toString()
```
Comment (TylerMatteo) [inline app/controllers/application.js:18]: This is a relatively small thing but I think it would be better to use the overload of this function that takes the name of a route and an "options" object with the queryParams, instead of "building" your own `URLSearchParams` object here. (See https://api.emberjs.com/ember/3.10/classes/Controller/methods/transitionToRoute?anchor=transitionToRoute)

File: app/controllers/application.js [hunk 2]
```diff
@@ -1,6 +1,24 @@
 import Controller from '@ember/controller';
 import { inject as service } from '@ember/service';
+import { action } from '@ember/object'
 
 export default class ApplicationController extends Controller {
   @service router;
+  
+  @action openExplorerURL() {
```
Comment (godfreyyeung) [inline app/controllers/application.js:8]: Nice code here to check sessionStorage and construct a url based on its values. 
An alternative you can explore is to move this code over to the route of the explorer page. 
This way, instead of defining this code at all the possible links to the Explorer page, you define it at the "gateway" of the Explorer page. So from whichever link or button the user clicks to get to the Explorer, this procedure to check sessionStorage still runs.

File: app/controllers/explorer.js [hunk 3]
```diff
@@ -38,7 +38,7 @@ export default class ExplorerController extends Controller {
   // The comparison geography ID.
   // Must match ID of one option in comparisonGeoOptions.
   // Default "0" maps to NYC.
-  @tracked compareTo = "0";
+  @tracked compareTo = "1";
```
Comment (TylerMatteo) [inline app/controllers/explorer.js:41]: Curious why this change? Don't we still want it to default to 0 so that the default comparison is citywide?

File: app/controllers/explorer.js [hunk 4]
```diff
@@ -121,7 +121,9 @@ export default class ExplorerController extends Controller {
   set source(newSource) {
     const { id } = newSource;
 
-    this.transitionToRoute('explorer', { queryParams: { source: id }});
+    sessionStorage.setItem('source', newSource.id)
```
Comment (TylerMatteo) [inline app/controllers/explorer.js:124]: It might be more trouble than it's worth but I think we should consider alternatives to using sessionStorage for this. Keeping track of "state" like this is the sort of thing we should use Ember for.

**Review by TylerMatteo — CHANGES_REQUESTED** [review]

File: app/controllers/application.js (hunk 2 above)
Comment (godfreyyeung) [reply app/controllers/application.js:8]: Here is the Explorer route file where you might check session storage: https://github.com/NYCPlanning/labs-factfinder/blob/develop/app/routes/explorer.js

I assume the main point of the ticket is to restore all the data table configurations like source, topic filters, Compare To, show reliability etc. So as long as loading session state in the route doesn't override the user's selected geoid and geography, it should be ok

File: app/controllers/application.js (hunk 1 above)
Comment (bmarchena) [reply app/controllers/application.js:18]: I originally tried using that approach, eg ```this.transitionToRoute('explorer', { queryParams })``` but it wouldn't load the page properly. It would take users to /explorer/**undefined**/**undefined**?compareTo=1 instead of /explorer/cities/NYC?compareTo=1 and I couldn't find a way to pass in those additional URL parameters

File: app/controllers/explorer.js (hunk 3 above)
Comment (bmarchena) [reply app/controllers/explorer.js:41]: In prod we're defaulting to Manhattan as the default comparison so I changed it here in for consistency. This initial state actually wasn't previously being used until you changed the comparison geography since the Explorer tab would link directly to /explorer/cities/NYC?**compareTo=1**

File: app/controllers/explorer.js (hunk 4 above)
Comment (bmarchena) [reply app/controllers/explorer.js:124]: I'm open to alternatives but from my research I didn't see many state management options that wouldn't significantly add to the scope of the ticket (but I'm also not super familiar with Ember yet so there may well be)

Comment (godfreyyeung) [reply app/controllers/explorer.js:124]: My thoughts on the options: 
1. Application controller - would not preserve state if you refresh the application
2. URLs - Preserves state across reloads and forward/backwards page navigation only if user strictly uses the same method to move between pages _both_ ways. Ie strictly uses application controls or strictly uses browser navigation. Requires you to add to URLs used for transitioning back to Map page. Cons: URL looks messy and awkward on the map page. 
3. Ember service - doesn't preserve state if you refresh the application
4. Local storage - preserves state if you refresh application. User can use a mix of browser and application navigation without worry of losing state. URL stays clean on the map page.

File: app/controllers/explorer.js (hunk 3 above)
Comment (TylerMatteo) [reply app/controllers/explorer.js:41]: Where do you see it defaulting to Manhattan? I picked a random census tract and clicked "View Data" in prod and it looks like it's comparing to citywide
image

Comment (TylerMatteo) [reply app/controllers/explorer.js:41]: oooh I see what's going on. In prod, we default the _comparison area_ to citywide if the user picks a geography on the map and then clicks "View Data" to reach the Explorer. If the user just clicks "Data Explorer" in the nav menu, they have citywide as their _selection_ and Manhattan as the comparison area. This change seems to make Manhattan the default comparison for users that get to the Explorer via "View Data", which I don't think we want. I think the logic here is the default comparison should be citywide, unless the selection is citywide, then its Manhattan

**Comment by TylerMatteo** [comment]
I think we're missing some user flows here. I pulled down your branch to do some testing and when I:
* Pick a census tract on the map
* Click "View Data" to go to the Explorer
* Switched to ACS Data and picked a different set of topics
* Clicked "Map" in the top nav
* Select a different tract
* Click View Data to go back to explorer
It seems to be going back to the default selection of 2020 Decennial and all topics

File: app/controllers/application.js (hunk 1 above)
Comment (TylerMatteo) [reply app/controllers/application.js:18]: The "cities" and "NYC" parts of "/explorer/cities/NYC" are [Dynamic Segments](https://guides.emberjs.com/v3.10.0/routing/defining-your-routes/#toc_dynamic-segments) so you have to specify the value for those as separate arguments when you use `transitionToRoute`. (note that dynamic segments aren't quite the same thing as query params but they serve a similar purpose). Try `this.transitionToRoute('explorer', 'cities', 'NYC', { queryParams });`

File: app/controllers/explorer.js (hunk 4 above)
Comment (TylerMatteo) [reply app/controllers/explorer.js:124]: @godfreyyeung thanks for laying those options out. Tbh I'm not sure if maintaining across application refreshes is something we want necessarily? Users might hit refresh specifically to "clear" state like this. I'm cool with eliminating a service as an option, I agree with the point you made earlier that the Ember docs seem suggest Services aren't intended to keep this kind of state.

File: app/controllers/application.js [hunk 5]
```diff
@@ -1,6 +1,24 @@
...
+    const explorerURL = "/explorer/cities/NYC"
+    let newURL;
+    if (sessionStorage.length) {
+      let queryParams = new URLSearchParams()
+
+      for (let key in sessionStorage) {
```
Comment (TylerMatteo) [inline app/controllers/application.js:14]: Looping through every key in sessionStorage like this is likely to introduce bugs down the road. `sessionStorage` is an instance of a class so it has more keys than just the ones you're setting. If you log out the `queryParams.toString()` in your action for navigating to the Explorer page, you'll see it includes params for other things we don't want, like `setItem=function`, `length=3`, etc. Apparently `transitionToRoute` is sanitizing those out but we should explicitly pull just the keys we expect to have and need from sessionStorage here.

File: app/controllers/explorer.js (hunk 4 above)
Comment (bmarchena) [reply app/controllers/explorer.js:124]: The issue is that in order for this to work we will need to preserve state across refreshes somehow, since clicking into any of the links in the nav bar does cause a refresh in the app

File: app/templates/application.hbs [hunk 6]
```diff
@@ -22,7 +22,7 @@
       {{/link-to}}
     </li>
     <li>
-      <a href="/explorer/cities/NYC?compareTo=1"
+      <a onclick={{action this.openExplorerURL}}
         class={{(if (eq this.router.currentRouteName "explorer") "active-route")}}
       >
```
Comment (bmarchena) [inline app/templates/application.hbs:27]: So here's something interesting – it looks like replacing this `a` tag with a `link-to` 'explorer' 'cities' 'NYC' causes the app to use the [queryParams we have bound](https://guides.emberjs.com/v3.4.0/routing/query-params/#toc_specifying-query-parameters) in the explorer [controller](https://github.com/NYCPlanning/labs-factfinder/blob/646f6bc3c9703e34b7f68797da1799b588e5d992/app/controllers/explorer.js#L17-L41). I'm willing to bet this is how the app was working previously. With this approach we don't have to use storage at all, but unfortunately the filters are only saved between the Map and Data Explorer tabs – hitting View Data will still lead users to the default Data Explorer view for that tract.

Comment (bmarchena) [reply app/templates/application.hbs:27]: cc @TylerMatteo @godfreyyeung

Comment (TylerMatteo) [reply app/templates/application.hbs]: Well that's promising, at least. I'm going to look into how the "View Data" button works because I'm pretty sure this isn't going to pass testing without that. I'm pretty sure users use that button much more than the actual "Data Explorer" tab.

**Review by TylerMatteo — APPROVED** [review]
Lgtm! I did some testing and it seemed to work with all the user paths I tried.
